roadmap/propagate-rust-panic-through-native-frame.md

```diff
@@ -46,3 +48,37 @@ cases (contains destructors, no affordances). We will also have to
 specify this on a per-target basis, as the details will vary depending
 on what exception mechanism is in use.
 
+Another consideration is whether the the native frames include a `catch` block.
+There are several possibilities here as well:
+
+* "catch native" -- e.g. a C++ block catching a specific subset of C++ exceptions
+* "catch all" -- e.g. a C++ catch without an exception argument
+* "catch and release" -- e.g. a C++ "catch all" that re-`throw`s the exception
+
+So far we do **not** have plans to specify these cases.
+
+## Unwind-termination and related concerns
+
+There are several possibilities for the top section of the diagram, above the
+native frames:
+
+* Rust frames - the `panic` re-enters Rust frames
+  * At this point, the `panic` will be no different from a normal `panic`,
+    unless and until it re-enters native frames.
+  * The `panic` may caught as usual, either with `catch_panic` or by crossing a
+    thread boundary.
+* Thread boundary - the exception propagates all the way to a native frame that
+  was invoked from another thread
+* App entry point - the unwind is not caught and does not cross a thread
+  boundary.
+  * This should cause the application to terminate.
+
+Again, these are all currenntly undefined.
+
+We **do** expect to make the first and last possibilities (re-entering Rust
+frames or reaching the app entry point) well-defined; their behavior should be
+a natural consequence of how the `"C unwind"` ABI is defined.
+
+We **do not** currently have plans to define the middle case, wherein a `panic`
+propagates through native frames all the way to a thread boundary.
+
```

Comment:
So this is a meta-comment about the whole file. I think this is the wrong approach to attack this problem.

The whole point of an ABI is to specify an interface. We have to specify whether Rust conforms to the interface or not, and if it does, how Rust behaves at its side of the interface. 

We could document how other native code specifies their behavior at their side of the interface, but we can't specify such behavior for that code, and we can't guarantee that such behavior does not change. 

Specifying here that if a foreign exception crosses a catch block in C++ the behavior is undefined is as meaningful as C++ specifying that integer arithmetic is undefined behavior in Rust. Sure, that's something that C++ could write in their language spec, but that does not make it true.

And for example, for all the cases considered here about what happens w/o destructors, catch blocks, thread boundaries, etc. C++ implementations actually specifies the behavior of foreign exceptions that adhere to the platform ABI, so the only way in which that could result in the UB specified here would be if the "C unwind" feature were unsound, e.g., and would not have the ABI that it says it does.

Reply:
So instead of writing what we think should be UB or not at the other side of the FFI, I'd very much prefer if we would just write what we instead _guarantee_ for our side, and what we guarantee for and require from the interface. These guarantees and requirements are what allow interfacing with native code correctly, since they tell this code what it can rely on, and what things it must uphold to avoid UB on our side. To avoid UB on its side, such code should refer to the specification of whatever language it is written in.

Since we have already said that functions with the `"C unwind"` ABI unwind with the platform ABI, Rust definitions with that ABI must uphold that guarantee, and so must functions that are declared in Rust with that ABI but defined in unknown code (independently of whether that code is Rust or not). Whether other native code can call those functions or not and what the semantics of that code are depend on that code. 

For the Rust side we need to define the behavior for the two cases:

* the unwinding is due to a Rust panic: in this case, it behaves like a normal Rust panic, for which we already have defined the behavior - anything else would break Rust->Rust unwinding through "C unwind".
* the unwinding is due to a "foreign" exception (e.g. a C++ `throw`): we can initially say that the behavior is undefined

And I think that would suffice for an MVP that allows Rust->C++->Rust to work correctly with strong stability guarantees for the targets that support the `"C unwind"` ABI (we should probably say somewhere that the `"C unwind"` feature is only available in some targets).

The undefined behavior caused by "foreign" exceptions can always be defined later.

Reply:
> C++ implementations actually specifies the behavior of foreign exceptions that adhere to the platform ABI...

This is conflating two kinds of specification. The C++ _standard_ specifies no such thing. What we are writing in these documents are more akin to a Rust _standard_ than to an _implementation_ specification. So it is just as reasonable for us to leave these things undefined as it is for C++ to do so.

We do want Rust _implementations_ supporting `extern "C unwind"` to specify their behavior, but this is a bit more of a long-term goal than simply omitting `nounwind`, which is why we have all of the future-facing verbiage, e.g. "we plan to define the behavior".

Reply:
> What we are writing in these documents are more akin to a Rust standard than to an implementation specification.

RFCs are not for a Rust standard decoupled from an implementation. They are for the Rust toolchain as a whole: the Rust reference, the Rust compiler, cargo, etc.

Reply:
@gnzlbg I'm not sure how true that is, but this isn't even an "RFC" right now.

Reply:
Isn't this material that should lead to an RFC at some point ?

Reply:
Yes, but for now let's just assume that the material in the roadmap is at "language specification" level.

Reply:
I'm going to mark this as "resolved", but I do think that ultimately the right way for us to proceed is going to specifying *what we guarantee as part of the "C unwind" ABI* -- and then discussing the implications of those guarantees